### ISSUE TITLE
[Feature Request] LAMMPS output directly into structured database (e.g. Hadoop on SQL)

### ISSUE THREAD

**Issue by ejmeitz** [lammps/lammps#3554]
**Summary**

It would be super useful if you could direct your LAMMPS output into some kind of database format (SQL etc.) instead of just text/YAML/JSON files.

**Detailed Description**

The LAMMPS output format is slowly getting more flexible (YAML added in Mar 2022) but I think a good next step would be to incorporate the ability for LAMMPS to write directly into a database. I do not know what kind of database would be best, but some kind of structured output that links the simulation details to the various outputs (thermo, traj etc.). Broadly, speaking the database would contain a structure to store per particle properties (like from a dump file or traj) and system level properties (like from a thermo file). It would also contain all the information necessary to re-run the simulation (in-file, structure etc.).

This would help facilitate the sharing and standardization of data and could also speed up data analysis as frameworks like Hadoop enable parallel access to data. It would also be possible to add a layer on top of this database that calculates common properties like RDF because the structure is consistent. This [paper](https://journalofbigdata.springeropen.com/articles/10.1186/s40537-014-0009-5#additional-information) has some suggested functions in Table 1 and a more detailed motivation for why this is important.

Thanks

**Comment by akohlmey** [comment]
@ejmeitz thank you for your suggestion. I am not so certain whether a direct integration of this kind of feature into LAMMPS would be as useful as you are anticipating. You need to keep in mind, that many LAMMPS calculations are run on HPC clusters and in parallel. Network traffic is usually not routed to the outside world from the compute nodes, so you would either have to run the database inside the cluster or set up some kind of network proxy/tunnel. Also, if the exchange with the database would not be asynchronously, you can add significant non-parallel overhead to the process.

With that in mind, a more effective approach - in my personal opinion - would be to have the database update run as an independent process, that can run on the part of the cluster with network access to outside computers. This can be done as a tool that reads/parses output files written by LAMMPS and then transmits the data to the database. If you configure a suitable LAMMPS dump style to create a single file for each output timestep, you would simply need a process that monitors a folder and processes all new files in that folder as they appear. That could probably be easily done in a scripting language like Python based on existing modules. Doing this directly in C++ may be significantly more complex. Also, you have the advantage of using independent software, so the tool could be adapted for other file formats and other simulation software packages.

**Comment by ejmeitz** [comment]
I definitely agree that a post-processing step to move the output into a database is easier. Especially, if you manipulate the LAMMPS output to be per-timestep like you mention. I guess what I was more after was data standardization and the ability to do SQL type queries to quickly calculate common analysis pipelines (RDF, heat capacity etc). A structured database just seemed like a good way to do that. Depending on the database used it could also speed up analysis (MapReduce) relative to parsing huge amounts of data through Python. I wonder if it would be possible to handle any LAMMPS output not just a specially formatted output. Due to the limited [output styles](https://docs.lammps.org/Howto_output.html) in LAMMPS this might be possible to do. I've been thinking about how to do this but thought I'd share the idea.

As for the direction integration...I guess I had not considered that many users do not have complete access to the cluster they run LAMMPS on. If you did have full access this idea is slightly less crazy.

**Comment by akohlmey** [comment]
People have been trying to establish some kind of "standard format" or "standard conventions" for MD simulation data for as long as I can remember (i.e. 25+ years). However, there is a lot of inertia:  there is already a well established environment of pre- and post-processing tools for almost all MD packages (with some overlap) and typical file formats are reasonably well supported. Getting everybody to agree to a new format *and* establishing it in commonly used tools for pre- and post-processing and analysis is a huge task. The people that typically would be using these tools are often content with spending as little effort as possible and settle for the lowest common denominator as they worry more about the results and the research to be conducted than the file formats and the complications that come with any of the many proposed standardizations. That is the main reason why minimal file formats like .xyz and .pdb are still so widely used despite their shortcomings.

A recently proposed standard is published at  https://www.openpmd.org/ (unfortunately the realization of that website is doing a good job of turning people away, so you may be better off checking out their github repo at https://github.com/openPMD/openPMD-standard )

> Depending on the database used it could also speed up analysis (MapReduce) relative to parsing huge amounts of data through Python.

Time consuming analysis is rarely limited by the time it takes to read the trajectory data. And even when reading the trajectory is consuming a lot of time the most time is spend on converting text to floating-point numbers, so storing data in binary formats (native LAMMPS, DCD, XTC, or NetCDF/HDF5 based). When you need to use parallelization, then another bottleneck is getting the data distributed again. In my experience, the parallelization offered by MapReduce will fail on anything where time correlation matters.  A lot of that processing can be done in LAMMPS directly during the simulation and in parallel, which is often preferable and most efficient.

> I wonder if it would be possible to handle any LAMMPS output

You have to provide some more explanation what you mean by that. Pretty much any data that LAMMPS produces *and* is worth keeping, can be directed to some supported output file format. As for any classical MD simulation, certain data simply cannot be stored because there would be so much of it and managing it would be a nightmare. It is usually simpler to just recompute it. Certain parts of LAMMPS itself are testament to the fact that it can be faster to compute the same thing multiple times (when running in parallel) rather than using algorithms that compute everything only once. GPU acceleration often works that way.

> many users do not have complete access to the cluster they run LAMMPS on

It is not so much a matter of having access, but a matter of how the networks on these clusters are configured. The kind of configuration needed to make a cluster work efficiently and manage it conveniently conflict with making the internal network of the cluster visible to the outside. In addition, even the login nodes are often well protected and firewalled so even pushing data from the login servers to somewhere outside can be difficult. People often underestimate the amount of automated hacking attempts that happen to *any* machine publicly visible on the internet and how easily security vulnerabilities can be automatically exploited. Running a database on the login nodes is often in conflict with the available storage capacity. Typically, HPC clusters are not set up for providing large amounts of permanent storage. Per-user storage often needs to be maintained on storage facilities that can be efficiently used from parallel calculations and with support for many concurrent accesses from different users. They also have to provide high data transfer rates, so that (big!) calculations can be checkpointed and restarted efficiently without wasting too much time.

In summary, there are lots of things that look good from far, but are far from good when looking closer, and the closer you look, the worse it gets.

**Comment by ejmeitz** [comment]
Thanks for the insight! The openPMD format does look interesting and check most of the boxes for what I was looking for. A database was just how my brain envisioned it but I think openPMD is simpler and does not have most of the issues you laid out.

Just curious, for things like openPMD what would it take in the community to get a feature like that added to LAMMPS? or is it more on the user to take LAMMPS data and put it in this format?

**Comment by akohlmey** [comment]
> Just curious, for things like openPMD what would it take in the community to get a feature like that added to LAMMPS?

For the most part it needs a volunteer that wants this enough to spend the time programming what needs to be programmed.

Please note that openPMD is a convention rather than a format. So you mostly would need to follow the naming of the convention. LAMMPS already has a *lot* of flexibility (almost all keywords for output properties can be renamed if you use the current development version), so it is mostly a question of writing a suitable input.

LAMMPS already has dump styles for two other conventions (based on netcdf and hdf5, respectively). You could use that as a starting point. Also the YAML format dump style has the most flexibility.

Similar as for moving output to a database, the simplest way to implement such a convention is to write a converter (ideally something that can work on-the-fly - as outlined - to reduce temporary storage requirements).  The hard job is to convince others to actually use some standard format. The harder job is to make people place their data where can can be accessed (after it was stored in some standard format). This requires to raise awareness for following even the most simple data management rules. In my personal experience, that awareness is very limited. In most cases, the PI designs the data management plan (because it is required in the proposal to get funding), but this information is rarely communicated to the grad students and postdocs and corresponding best practices are thus rarely followed.
